validator_client/src/duties_service.rs

```diff
@@ -575,7 +645,17 @@ async fn poll_beacon_attesters<T: SlotClock + 'static, E: EthSpec>(
         metrics::start_timer_vec(&metrics::DUTIES_SERVICE_TIMES, &[metrics::SUBSCRIPTIONS]);
 
     // This vector is likely to be a little oversized, but it won't reallocate.
-    let mut subscriptions = Vec::with_capacity(local_pubkeys.len() * 2);
+    // The calculation is based on the fact that every validator has
+    // `ATTESTATION_SUBSCRIPTION_OFFSETS.len()` subscriptions to send each epoch. We assume these
+    // subscriptions are distributed roughly evenly across the slots, but include an extra margin of
+    // 1/4 to hedge against imbalances.
+    let num_expected_subscriptions = std::cmp::max(
+        1,
+        5 * local_pubkeys.len() * ATTESTATION_SUBSCRIPTION_OFFSETS.len()
+            / (4 * E::slots_per_epoch() as usize),
```

Comment:
Not really that important, but I didn't follow this logic. Why 5x keys?  Also the division by 4? 
Naively i would expect 50% of keys to do `ATTESTATION_SUBSCRIPTION_OFFSET.len() -1` and they lie in one half of the epoch. 

It doesn't really matter tho, we're just trying to save some allocation, the larger the better, probably

Reply:
The 5/4 is so the vec is 1/4 larger than it should be.

Other than the 5/4 the bound is quite tight because we are looking at 2 epochs total (current and next), there are `2 * local_pubkeys.len()` attestation duties, and each duty gets sent at `ATTESTATION_SUBSCRIPTION_OFFSETS.len()` slots. So there are `2 * local_pubkeys.len() * ATTESTATION_SUBSCRIPTION_OFFSETS.len()` subscription-slots to be distributed. They're approximately distributed over 2 epochs (current and next) but with a slight offset from the epoch start, e.g. some of the subscription-slots for the current epoch were in the previous epoch, but this is balanced out by some of the subscription-slots for the `N + 2` epoch being in the next epoch (we just don't know it yet). Therefore the total number of subscription-slots per slot is: `(2 * local_pubkeys.len() * ATTESTATION_SUBSCRIPTION_OFFSETS.len()) / (2 * E::slots_per_epoch())`, and we can cancel the factor of 2.

Experimentally the number of subscriptions per slot is very close to this. For 10k validators, we'd expect 2.5k subscriptions without the 5/4 factor, and in practice we're very close to this:

> DEBG Sent attestation subscriptions expected: 3124, count: 2512
> DEBG Sent attestation subscriptions expected: 3124, count: 2560
> DEBG Sent attestation subscriptions expected: 3124, count: 2535
> DEBG Sent attestation subscriptions expected: 3124, count: 2483